### ISSUE TITLE
Custom sampler for Seq2Seq models to avoid padding

### ISSUE THREAD

**Issue by chrisvdweth** [pytorch/pytorch#25743]
## 🚀 Feature

As proposed in this [PyTorch forum post](https://discuss.pytorch.org/t/using-batches-for-seq2seq-models/55205): Custom sampler for Sequence-to-Sequence models (but also for sequence classification, sequence tagging, autoencoders, etc.) that forms homogeneous batches, i.e., in the same batch all input sequences have the same length, and all target sequences have the same length. For example a batch might contain 32 pairs with all input sequences having length 12 and all target sequences having length 15.

## Motivation

Most Sequence-to-Sequence tutorials assume batch sizes of 1, since target sequences typically and with a end-of-sequence (EOS) token without padding. Thus, random batches of larger would include target sequences with different length. Batches of size 1, of course, show poor performance w.r.t. to training time compared the larger batches

While padding of input sequences is common (at least for sequence classification), it's effects on the accuracy are not always clear. Optional packing using `PackedSequence` arguably mitigates this but also adds noticeable overhead to the performance of the training (anecdotal bservation: up 10% longer training)

## Pitch

A custom sampler organizes all data items, here: pairs of sequences, into batch, so that each batch contains only pairs with equal inputs lengths and equal output lengths. For example a batch might contain 32 pairs with all input sequences having length 12 and all target sequences having length 15. The benefits are:

* Obviously, larger batches that yield better training times compared to batches of size 1
* No padding and optional packing need (potentially avoid all involved side effects)

## Alternatives

`torchtext` has its own `BucketIterator` that organizes batches to minimize padding (not completely avoiding it). 

## Additional context

To see how the combination of input and target lengths are distributed, I’ve made a test with a real-world dataset for machine translation. I’ve set the batch size to 32. The figure below shows the distribution of the sizes of all batches. As one can see, the vast majority of batches is indeed full (i.e., 32 sequence pairs). This shouldn’t really be surprising since:

* Batch sizes (e.g., 32 or 64) are is essentially nothing given large datasets of millions of sequences pairs or more. Thus, the chance that enough pairs share the same input and target lengths is high.
* The combination of input and target lengths is typically not independent. For example, an input of length 5 generally does not have a target length of 20 but in a similar range.

![bucketbatchsample-batch-size-distribution](https://user-images.githubusercontent.com/22168190/64392082-c79dae80-d07d-11e9-8a5a-9be614350457.png)

**Comment by soumith** [comment]
cc: @cpuhrsch @zhangguanheng66

**Comment by fmassa** [comment]
What you are asking for can probably be achieved with GroupedBatchSampler from https://github.com/pytorch/vision/blob/master/references/detection/group_by_aspect_ratio.py
Where you pass as groups the lengths of your sentences.

**Comment by zhangguanheng66** [comment]
The `BucketIterator` sampler in `torchtext` is able to sort the sentences by lengths. As Francisco mentioned, it should be achievable with `GroupedBatchSampler`. @fmassa Do you think we should move it to `torch.data.utils` for all the domains? I guess it may be useful for some audio data @vincentqb .

**Comment by fmassa** [comment]
> Do you think we should move it to torch.data.utils for all the domains? I guess it may be useful for some audio data 

I think it can be useful in general, yes, so it might make sense to move it (or an improved version of it) to `torch.utils.data`.

**Comment by vincentqb** [comment]
I don't immediately have an example for torchaudio, but if the two of you already have use cases, it makes sense to me to move a version of `GroupedBatchSampler` into `torch.utils.data`.

**Comment by zhangguanheng66** [comment]
It's quite common for the text sequences. For some NLP tasks, we would like to batch those sequences with similar lengths to avoid unnecessary padding.
